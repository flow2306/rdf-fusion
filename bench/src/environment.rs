use crate::BenchmarkingOptions;
use crate::benchmarks::BenchmarkName;
use crate::prepare::{PrepRequirement, prepare_run_closure, prepare_run_command};
use crate::prepare::{ensure_file_download, prepare_file_download};
use anyhow::bail;
use datafusion::execution::runtime_env::{RuntimeEnv, RuntimeEnvBuilder};
use datafusion::prelude::SessionConfig;
use rdf_fusion::store::Store;
use std::path::{Path, PathBuf};
use std::sync::{Arc, Mutex};

/// Represents a context used to execute benchmarks.
pub struct RdfFusionBenchContext {
    /// General options for the benchmarks.
    options: BenchmarkingOptions,
    /// The path to the data dir.
    data_dir: Mutex<PathBuf>,
    /// The path to the results dir.
    results_dir: Mutex<PathBuf>,
}

impl RdfFusionBenchContext {
    /// Creates a new [RdfFusionBenchContext].
    pub fn new(
        options: BenchmarkingOptions,
        data_dir: PathBuf,
        results_dir: PathBuf,
    ) -> Self {
        Self {
            options,
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(results_dir),
        }
    }

    /// Creates a new [RdfFusionBenchContext] used in the criterion benchmarks.
    pub fn new_for_criterion(data_dir: PathBuf, target_partitions: usize) -> Self {
        Self {
            options: BenchmarkingOptions {
                verbose_results: false,
                target_partitions: Some(target_partitions),
                memory_size: None,
            },
            data_dir: Mutex::new(data_dir),
            results_dir: Mutex::new(PathBuf::from("/temp")),
        }
    }

    /// Returns the [BenchmarkingOptions] for this context.
    pub fn options(&self) -> &BenchmarkingOptions {
        &self.options
    }

    /// Resolves a relative path `file` against the data directory.
    pub fn join_data_dir(&self, file: &Path) -> anyhow::Result<PathBuf> {
        if !file.is_relative() {
            bail!("Only relative paths can be resolved.")
        }

        Ok(self.data_dir.lock().unwrap().join(file))
    }

    pub fn create_store(&self) -> Store {
        let mut config = SessionConfig::new()
            .with_batch_size(8192)
            .with_target_partitions(self.options.target_partitions.unwrap_or(1));

        let options = config.options_mut();
        options.optimizer.enable_dynamic_filter_pushdown = true;

        let runtime_enc = match self.options.memory_size {
            None => Arc::new(RuntimeEnv::default()),
            Some(memory_size) => RuntimeEnvBuilder::new()
                .with_memory_limit(memory_size * 1024 * 1024, 1f64)
                .build_arc()
                .expect("Only setting memory limit"),
        };
        Store::new_with_datafusion_config(config, runtime_enc)
    }

    /// Creates a new folder in the results directory and uses it until [Self::pop_dir] is
    /// called.
    ///
    /// This can be used to create folder hierarchies to separate the results of different
    /// benchmarks.
    #[allow(clippy::create_dir)]
    #[allow(clippy::unwrap_used, reason = "Mutex poisoning")]
    pub fn push_dir(
        &self,
        data_dir_name: &str,
        results_dir_name: &str,
    ) -> anyhow::Result<()> {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.push(data_dir_name);
        results_dir.push(results_dir_name);

        Ok(())
    }

    /// Pops the last directory from the stack.
    pub fn pop_dir(&self) {
        let mut data_dir = self.data_dir.lock().unwrap();
        let mut results_dir = self.results_dir.lock().unwrap();

        data_dir.pop();
        results_dir.pop();
    }

    /// Creates a new bencher and modifies the context for this benchmark.
    pub fn create_benchmark_context(
        &self,
        benchmark_name: BenchmarkName,
    ) -> anyhow::Result<BenchmarkContext<'_>> {
        self.push_dir(
            &benchmark_name.data_dir_name(),
            &benchmark_name.results_dir_name(),
        )?;
        Ok(BenchmarkContext {
            context: self,
            benchmark_name,
        })
    }
}

/// A benchmarker that can be used to execute benchmarks.
///
/// It holds a reference to the current context to store its results.
pub struct BenchmarkContext<'ctx> {
    /// Reference to the benchmarking context.
    context: &'ctx RdfFusionBenchContext,
    /// Name of the benchmark that is being executed.
    benchmark_name: BenchmarkName,
}

impl<'ctx> BenchmarkContext<'ctx> {
    /// Returns a reference to the benchmarking context.
    pub fn parent(&self) -> &RdfFusionBenchContext {
        self.context
    }

    /// Returns the name of the benchmark that is being executed.
    pub fn benchmark_name(&self) -> BenchmarkName {
        self.benchmark_name
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn data_dir(&self) -> PathBuf {
        self.context.data_dir.lock().unwrap().clone()
    }

    /// Returns the path to the results directory of this benchmark.
    pub fn results_dir(&self) -> PathBuf {
        self.context.results_dir.lock().unwrap().clone()
    }

    /// Prepares the context such that `requirement` is fulfilled.
    pub async fn prepare_requirement(
        &self,
        requirement: PrepRequirement,
    ) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload {
                url,
                file_name,
                action,
            } => prepare_file_download(self, url, file_name, action).await,
            PrepRequirement::RunClosure { execute, .. } => {
                prepare_run_closure(self, &execute)
            }
            PrepRequirement::RunCommand {
                workdir,
                program,
                args,
                ..
            } => {
                let workdir = self.context.join_data_dir(&workdir)?;
                prepare_run_command(&workdir, &program, &args)
            }
        }
    }

    /// Ensures that the `requirement` is fulfilled in this context.
    pub fn ensure_requirement(&self, requirement: PrepRequirement) -> anyhow::Result<()> {
        match requirement {
            PrepRequirement::FileDownload { file_name, .. } => {
                ensure_file_download(self, file_name.as_path())
            }
            PrepRequirement::RunClosure {
                check_requirement, ..
            }
            | PrepRequirement::RunCommand {
                check_requirement, ..
            } => check_requirement(self),
        }
    }
}

/// Pops the results directory from the context when the bencher is dropped.
impl Drop for BenchmarkContext<'_> {
    fn drop(&mut self) {
        self.context.pop_dir();
    }
}
