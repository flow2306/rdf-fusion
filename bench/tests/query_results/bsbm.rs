//! This file contains tests for an adapte version of the BSBM explore use case. The adaption have
//! been made such that the queries produce stable results.
//!
//! The results have been compared against GraphDB 11.0.1. The following curl command can be used to
//! get the results of GraphDB. Ideally, one can pipe the result of the request into a file or
//! directly to the clipboard using `| xclip -selection clipboard` (or similar). You can also
//! download the JSON result from the GraphDB UI.
//!
//! ```bash
//! curl -X POST \
//!     "http://<graphdb_url>/repositories/bsbm" \
//!     -H "Content-Type: application/sparql-query" \
//!     -H "Accept: application/sparql-results+json" \
//!     --data-binary '<query>'
//! ```
//!
//! Then, even though we pretty-print our results, there will be some differences (e.g., spacing,
//! order of keys). You can use a tool that semantically compares JSON files to "quickly" check
//! the results of a new test (e.g., [JSON Compare](https://jsoncompare.org/)). `CONSTRUCT` queries
//! have been compared manually.

use crate::query_results::{run_graph_result_query, run_select_query};
use insta::assert_snapshot;
use rdf_fusion_bench::benchmarks::Benchmark;
use rdf_fusion_bench::benchmarks::bsbm::{BsbmBenchmark, ExploreUseCase, NumProducts};
use rdf_fusion_bench::environment::RdfFusionBenchContext;
use std::path::PathBuf;

#[tokio::test]
pub async fn bsbm_1000_test_results() {
    let benchmarking_context =
        RdfFusionBenchContext::new_for_criterion(PathBuf::from("./data"), 1);
    let benchmark =
        BsbmBenchmark::<ExploreUseCase>::try_new(NumProducts::N1_000, None).unwrap();
    let benchmark_name = benchmark.name();
    let ctx = benchmarking_context
        .create_benchmark_context(benchmark_name)
        .unwrap();

    let store = benchmark.prepare_store(&ctx).await.unwrap();

    //
    // Explore
    //

    assert_snapshot!(
        "Explore Q1",
        run_select_query(&store, include_str!("./queries/explore-q1.sparql")).await
    );
    assert_snapshot!(
        "Explore Q2 (empty optional)",
        run_select_query(
            &store,
            include_str!("./queries/explore-q2-empty-optional.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Explore Q2 (non-empty optional)",
        run_select_query(
            &store,
            include_str!("./queries/explore-q2-non-empty-optional.sparql")
        )
        .await
    );
    assert_snapshot!(
        "Explore Q3",
        run_select_query(&store, include_str!("./queries/explore-q3.sparql")).await
    );
    assert_snapshot!(
        "Explore Q4",
        run_select_query(&store, include_str!("./queries/explore-q4.sparql")).await
    );
    assert_snapshot!(
        "Explore Q5",
        run_select_query(&store, include_str!("./queries/explore-q5.sparql")).await
    );
    assert_snapshot!(
        "Explore Q7",
        run_select_query(&store, include_str!("./queries/explore-q7.sparql")).await
    );
    assert_snapshot!(
        "Explore Q8",
        run_select_query(&store, include_str!("./queries/explore-q8.sparql")).await
    );
    assert_snapshot!(
        "Explore Q9",
        run_graph_result_query(&store, include_str!("./queries/explore-q9.sparql")).await
    );
    assert_snapshot!(
        "Explore Q10",
        run_select_query(&store, include_str!("./queries/explore-q10.sparql")).await
    );
    assert_snapshot!(
        "Explore Q11",
        run_select_query(&store, include_str!("./queries/explore-q11.sparql")).await
    );
    assert_snapshot!(
        "Explore Q12",
        run_graph_result_query(&store, include_str!("./queries/explore-q12.sparql"))
            .await
    );

    //
    // Business Intelligence
    //

    assert_snapshot!(
        "Business Intelligence Q1",
        run_select_query(&store, include_str!("./queries/bi-q1.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q2",
        run_select_query(&store, include_str!("./queries/bi-q2.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q3",
        run_select_query(&store, include_str!("./queries/bi-q3.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q4",
        run_select_query(&store, include_str!("./queries/bi-q4.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q5",
        run_select_query(&store, include_str!("./queries/bi-q5.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q6",
        run_select_query(&store, include_str!("./queries/bi-q6.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q7",
        run_select_query(&store, include_str!("./queries/bi-q7.sparql")).await
    );
    assert_snapshot!(
        "Business Intelligence Q8",
        run_select_query(&store, include_str!("./queries/bi-q8.sparql")).await
    );
}
