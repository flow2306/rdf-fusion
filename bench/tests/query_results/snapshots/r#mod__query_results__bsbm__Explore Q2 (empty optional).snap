---
source: bench/tests/query_results/bsbm.rs
expression: "run_select_query(&store,\ninclude_str!(\"./queries/explore-q2-empty-optional.sparql\")).await"
---
{
  "head": {
    "vars": [
      "label",
      "comment",
      "producer",
      "productFeature",
      "propertyTextual1",
      "propertyTextual2",
      "propertyTextual3",
      "propertyNumeric1",
      "propertyNumeric2",
      "propertyTextual4",
      "propertyTextual5",
      "propertyNumeric4"
    ]
  },
  "results": {
    "bindings": [
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "adjustment"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "exteriorizing pannier tsunamic"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "graciousness flunkeys knave"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "grudges"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "intrusions tantalizers hesitating"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "orthoepist brats"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "paralyzed"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "sistrums"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      },
      {
        "comment": {
          "type": "literal",
          "value": "unlearn cowshed baldhead secularism transferror chastiser bries fiducially coacted unreplaced hockeys overachieving septettes luteum mooter benefactrix analemmas demises sponsoring homesite clairvoyance tonuses students unescapably twinship revisionist identifers sapphisms folderol finalization saleswomen vivants bookrack hairlines prickles besot unwaveringly headlights waterborne crosspatches bouts hunting isomerize monkeyshine delfts subpoenas boob praised villainies beguines darvon weirdies midlines estimations lacerable postfaces restamping occasionally opiates ticktock bugbane ineptness hydrofoils hissing treatment excommunicated fistulas corotate grooves legitimist corruptness fiefs paramedical nodding misidentifying chromaticity triced florentines expatriated hatefulness claps grapefruits ensure dissipates shelling pawns hangnails gills pantywaist reimprisoning blah announcer unhorse pointy compositions tens regerminated bushwack expansions vernacularly sarcasms remeasuring lineate polypous rewritten backslapping segmentary vaunters callosities airier fortifying oversides"
        },
        "label": {
          "type": "literal",
          "value": "vesseled venose"
        },
        "producer": {
          "type": "literal",
          "value": "cashmeres prearranges"
        },
        "productFeature": {
          "type": "literal",
          "value": "stewing existence countermands"
        },
        "propertyNumeric1": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "432"
        },
        "propertyNumeric2": {
          "datatype": "http://www.w3.org/2001/XMLSchema#integer",
          "type": "literal",
          "value": "1646"
        },
        "propertyTextual1": {
          "type": "literal",
          "value": "vacillated alienated cardiometer tittle chromizes obsoleting mahout warrens bestirred quadruplicating"
        },
        "propertyTextual2": {
          "type": "literal",
          "value": "bladed dobbins conjuring restorers mopiest backslide instituting finer mythos monotheism chewable etatism"
        },
        "propertyTextual3": {
          "type": "literal",
          "value": "massagist softballs collies"
        },
        "propertyTextual5": {
          "type": "literal",
          "value": "polyvinyl legalized penthouses expires"
        }
      }
    ]
  }
}
